pallets/phala/src/stakepool.rs

```diff
@@ -45,8 +99,80 @@ pub mod pallet {
 	impl<T: Config> Pallet<T> {
 		/// Creates a new stake pool
 		#[pallet::weight(0)]
-		pub fn create(origin: OriginFor<T>, id: u64) -> DispatchResult {
-			panic!("unimplemented")
+		pub fn create(origin: OriginFor<T>) -> DispatchResult {
+			let owner = ensure_signed(origin)?;
+
+			let pid = PoolCount::<T>::get();
+			PoolCount::<T>::mutate(|id| *id = pid + 1);
+			MiningPools::<T>::insert(
+				pid + 1,
+				PoolInfo {
+					pid: pid + 1,
+					owner: owner.clone(),
+					state: PoolState::default(),
+					payout_commission: Zero::zero(),
+					owner_reward: Zero::zero(),
+					pool_acc: Zero::zero(),
+					last_reward_block: T::BlockNumber::zero(),
+					total_staked: Zero::zero(),
+					extra_staked: Zero::zero(),
+					workers: vec![],
+				},
+			);
+			Self::deposit_event(Event::<T>::PoolCreated(owner, pid));
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn add_worker(
+			origin: OriginFor<T>,
+			pid: u64,
+			pubkey: WorkerPublicKey,
+		) -> DispatchResult {
+			let owner = ensure_signed(origin)?;
+
+			// make sure the worker has registered
+			ensure!(
+				registry::Worker::<T>::contains_key(&pubkey),
+				Error::<T>::WorkerNotRegistered
+			);
+			// check the wheather the owner was bounded as operator
+			let worker_info = registry::Worker::<T>::get(&pubkey).unwrap();
+			ensure!(
+				worker_info.operator == Some(owner.clone()),
+				Error::<T>::UnauthorizedOperator
+			);
+
+			// origin must be owner of pool
+			let mut pool_info = Self::mining_pools(pid).ok_or(Error::<T>::PoolNotExist)?;
+			ensure!(
+				pool_info.owner == owner.clone(),
+				Error::<T>::UnauthorizedPoolOwner
+			);
+			// make sure worker has not been not added
+			let mut workers = pool_info.workers;
+			ensure!(workers.contains(&pubkey), Error::<T>::WorkerHasAdded);
+
+			// generate miner account
+			let miner: T::AccountId =
+				STAKEPOOL_PALLETID.into_sub_account((pid, owner, pubkey.clone()).encode());
```

Comment:
https://github.com/paritytech/substrate/blob/d22f0df5ec001699f85e5be8b1a00d333b15de25/primitives/runtime/src/traits.rs#L1145-L1164

Subaccount is not a choice. It basically just take the first 32 bytes (appending zeros if less) to make a new AccountId32. In our case, the argument is even longer than 32 bytes. It will take 8 bytes pid + 24 bytes owner address. Then the worker pubkey is omitted.

We should hash them instead, e.g.:

```
blake2b_256((STAKEPOOL_PALLETID, pid, owner, pubkey.clone()).encode())
```